File: PanoramicDataWin8/controller/data/sim/DataBinner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using System.Dynamic;
using System.Diagnostics;
using PanoramicDataWin8.utils;
using System.Runtime.Serialization;
using System.IO;
using System.Runtime.Serialization.Json;
using Newtonsoft.Json;
using System.Collections.ObjectModel;
using Newtonsoft.Json.Serialization;
using Windows.ApplicationModel.Core;
using Windows.UI.Core;
using PanoramicDataWin8.model.data;
using PanoramicDataWin8.model.data.common;

namespace PanoramicDataWin8.controller.data.sim
{
    public class DataBinner
    {
        public bool Incremental { get; set; }
        public List<AxisType> AxisTypes { get; set; }
        public List<bool> IsAxisAggregated { get; set; }
        public List<double> NrOfBins { get; set; }
        public BinStructure BinStructure { get; set; }
        public List<InputFieldModel> Dimensions { get; set; }

        public void BinStep(List<DataRow> dataRows)
        {
            List<double> dataMins = new List<double>();
            List<double> dataMaxs = new List<double>();
            List<BinRange> binRanges = null;

            if (dataRows.Count == 0)
            {
                return;
            }

            getMinMaxValuesFromSamples(dataRows, dataMins, dataMaxs);

            if (BinStructure != null)
            {
                for (int d = 0; d < Dimensions.Count; d++)
                {
                    dataMins[d] = Math.Min(BinStructure.BinRanges[d].DataMinValue, dataMins[d]);
                    dataMaxs[d] = Math.Max(BinStructure.BinRanges[d].DataMaxValue, dataMaxs[d]);
                }
                binRanges = BinStructure.BinRanges.ToArray().ToList();
            }
            else
            {
                binRanges = createBinRanges(dataMins, dataMaxs);
                BinStructure = initializeBinStructure(binRanges);
            }
            for (int i = 0; i < BinStructure.BinRanges.Count; i++)
            {
                binRanges[i] = BinStructure.BinRanges[i].GetUpdatedBinRange(dataMins[i], dataMaxs[i]);
            }

            BinStructure tempBinStructure = initializeBinStructure(binRanges);
            binSamples(tempBinStructure, dataRows);

            // re-map old bins
            tempBinStructure.Map(BinStructure);
            BinStructure = tempBinStructure;
        }

        private List<BinRange> createBinRanges(List<double> dataMins, List<double> dataMax)
        {
            List<BinRange> binRanges = new List<BinRange>();
            for (int d = 0; d < Dimensions.Count; d++)
            {
                BinRange scale = null;
                if (IsAxisAggregated[d])
                {
                    scale = AggregateBinRange.Initialize();
                }
                else if (AxisTypes[d] == AxisType.Time || AxisTypes[d] == AxisType.Date)
                {
                    scale = DateTimeBinRange.Initialize(dataMins[d], dataMax[d], NrOfBins[d]);
                }
                else if (AxisTypes[d] == AxisType.Quantitative)
                {
                    scale = QuantitativeBinRange.Initialize(dataMins[d], dataMax[d], NrOfBins[d], Dimensions[d].InputDataType == InputDataTypeConstants.INT);
                }
                else
                {
                    scale = NominalBinRange.Initialize(dataMins[d], dataMax[d], NrOfBins[d]);
                }
                binRanges.Add(scale);
            }
            return binRanges;
        }

        private void getMinMaxValuesFromSamples(List<DataRow> dataRows, List<double> dataMins, List<double> dataMaxs)
        {
            for (int d = 0; d < Dimensions.Count; d++)
            {
                var dimension = Dimensions[d];
                var dataMin = dataRows.Select(dp => dp.VisualizationValues[dimension]).Where(dp => dp.HasValue).Min(dp => dp.Value);
                var dataMax = dataRows.Select(dp => dp.VisualizationValues[dimension]).Where(dp => dp.HasValue).Max(dp => dp.Value);

                if (dataMax == dataMin)
                {
                    if (AxisTypes[d] != AxisType.Quantitative)
                    {
                        dataMin -= 0;
                        dataMax += 0.1;
                    }
                    else
                    {
                        dataMin -= 1;
                        dataMax += 1;
                    }
                }
                dataMins.Add(dataMin);
                dataMaxs.Add(dataMax);
            }
        }
        
        private void binSamples(BinStructure binStructure, List<DataRow> samples)
        {
            foreach (var sample in samples)
            {
                BinIndex binIndex = new BinIndex();

                for (int d = 0; d < binStructure.BinRanges.Count; d++)
                {
                    double? value = sample.VisualizationValues[Dimensions[d]];
                    binIndex.Indices.Add(value.HasValue ? binStructure.BinRanges[d].GetIndex(value.Value) : -1);
                }
                Bin bin = null;
                if (binStructure.Bins.TryGetValue(binIndex, out bin))
                {
                    bin.Samples.Add(sample);
                }
                else
                {
                    binStructure.NullCount++;
                }
            }
        }

        private BinStructure initializeBinStructure(List<BinRange> binRanges)
        {
            BinStructure binStructure = new BinStructure();
            binStructure.BinRanges = binRanges;

            Dictionary<BinIndex, Bin> bins = new Dictionary<BinIndex, Bin>();
            recursiveCreateBins(binRanges.ToArray().ToList(), bins, new List<Span>());
            binStructure.Bins = bins;

            return binStructure;
        }

        private void recursiveCreateBins(List<BinRange> previousBinRangesLeft, Dictionary<BinIndex, Bin> bins, List<Span> previousSpans)
        {
            List<BinRange> binRangesLeft = new List<BinRange>();
            binRangesLeft.AddRange(previousBinRangesLeft);

            BinRange binRange = binRangesLeft[0];
            binRangesLeft.RemoveAt(0);

            foreach (double x in binRange.GetBins())
            {
                Span span = new Span() { Min = x, Max = binRange.AddStep(x), Index = binRange.GetIndex(x) };
                List<Span> spans = new List<Span>();
                spans.AddRange(previousSpans);
                spans.Add(span);
                if (binRangesLeft.Count == 0)
                {
                    Bin bin = new Bin()
                    {
                        Spans = spans,
                        BinIndex = new BinIndex(spans.Select(s => s.Index).ToArray())
                    };
                    bins.Add(bin.BinIndex, bin);
                }
                else
                {
                    recursiveCreateBins(binRangesLeft, bins, spans);
                }
            }
        }
    }
}

File: backend/databinner.py
#!/usr/bin/python
import json
import time
import numpy as np
import pandas as pd
import math
from binrange import *
import sys

class DataBinner():

    def __init__(self, dimensions, dimensionAggregateFunctions, nrOfBins, toAggregate, toAggregateFunctions, brushes):
        self.dimensions = dimensions
        self.dimensionAggregateFunctions = dimensionAggregateFunctions
        self.toAggregate = toAggregate
        self.toAggregateFunctions = toAggregateFunctions
        self.nrOfBins = nrOfBins
        self.brushes = brushes
        self.binStructure = None

    def bin(self, df, progress):
        dataMins = [df[dimension].min() for dimension in self.dimensions]
        dataMaxs  = [df[dimension].max() for dimension in self.dimensions]
        binRanges = []
        
        if not self.binStructure is None:
            binRanges = self.binStructure.binRanges
            for idx, val in enumerate(self.dimensions):
                dataMins[idx] = min(binRanges[idx].dataMinValue, dataMins[idx])
                dataMaxs[idx] = max(binRanges[idx].dataMaxValue, dataMaxs[idx])
        else:
            binRanges = self.__createBinRanges(dataMins, dataMaxs, df)
            self.binStructure = self.__initializeBinStructure(binRanges)
            
        for idx, val in enumerate(self.binStructure.binRanges):
            self.binStructure.binRanges[idx] = self.binStructure.binRanges[idx].getUpdatedBinRange(dataMins[idx], dataMaxs[idx], df, self.dimensions[idx])
        
        tempBinStructure = self.__initializeBinStructure(binRanges)
        tempBinStructure.map(self.binStructure)
        self.binStructure = tempBinStructure
        self.__binSamples(df, progress)

    def __binSamples(self, df, progress):
        #b1, b2, overlap, all
        currentBrushes = range(0, len(self.brushes) + 2)
        self.binStructure.initializeBatchAggregation(self.dimensions, self.toAggregate, self.toAggregateFunctions, currentBrushes, df)
    
        brushDataFrames = []
        mt = pd.Series([False] * len(df), index=df.index)
        fu = pd.Series([True] * len(df), index=df.index)
        if len(self.brushes) > 0:
            overlapMask = fu
            brushMasks = []
            
            restMask = fu
            for brush in self.brushes:
                brushMask = fu
                if brush != '':
                    brushMask = df.eval(brush)
                brushMasks.append(brushMask)
                overlapMask = (overlapMask & brushMask)                
                restMask = (restMask & ~brushMask)
            
            if len(self.brushes) <= 1:
                overlapMask = mt
            for brushMask in brushMasks:
                brushDataFrames.append(df[(brushMask & ~overlapMask)])

            brushDataFrames.append(df[overlapMask])
            brushDataFrames.append(df[restMask])
        else:
            brushDataFrames = [df[mt], df]
        
        start = time.time()
        for brushIndex, brushDataFrame in enumerate(brushDataFrames):
            toAggregateBrushes = [brushIndex, currentBrushes[-1]]
            dfDTypes = brushDataFrame.dtypes
            
            if len(brushDataFrame):
                #brushDataFrame['binIndex'] = brushDataFrame.apply(lambda x: self.__createAggregateKey(x), axis=1).astype(object)
                #grouped = brushDataFrame.groupby('binIndex', axis=1)
                #print grouped.first()
                #print grouped.apply(lambda x: len(x), axis=1)
                brushDataFrame.apply(lambda x: self.__runAggregation(x, dfDTypes, toAggregateBrushes, currentBrushes, progress, brushDataFrame), axis=1).astype(object)
                
        end = time.time()
        print (end - start)
        
        self.binStructure.endBatchAggregation(self.dimensions, self.toAggregate, self.toAggregateFunctions, currentBrushes, df, progress)
    
    def __runAggregation(self, row, dTypes, brushes, currentBrushes, progress, brushDataFrame):
        binIndex = []
        for idx, binRange in enumerate(self.binStructure.binRanges):
            value = row[self.dimensions[idx]]
            if isinstance(binRange, NominalBinRange):
                binIndex.append(self.binStructure.binRanges[idx].getIndexFromValue(value))    
            else:
                binIndex.append(self.binStructure.binRanges[idx].getIndex(value))        
        
        maxCount = sys.float_info.max;
        #bin = self.binStructure.bins[row['binIndex'].index]
        bin = self.binStructure.bins[tuple(binIndex)]
        bin.count += 1
        for brushIndex in brushes:
            for idx, aggregate in enumerate(self.toAggregate):
                aggregateFunction = self.toAggregateFunctions[idx]
                aggregateKey = (aggregate, aggregateFunction, brushIndex)
                dType = dTypes[aggregate]
                
                currentValue = bin.values[aggregateKey]
                value = row[aggregate]
                
                if aggregateFunction == 'Max':
                    currentValue = max(currentValue, value)
                    
                elif aggregateFunction == 'Min':
                    currentValue = min(currentValue, value)    
                    
                elif aggregateFunction == 'Sum':
                    currentValue = currentValue + value
                    
                elif aggregateFunction == 'Avg':
                    currentValue = ((currentValue * bin.counts[aggregateKey]) + value) / (bin.counts[aggregateKey] + 1)
                    n = bin.ns[aggregateKey] + 1
                    bin.ns[aggregateKey] = n
                    x = value
                    bin.means[aggregateKey] = bin.means[aggregateKey] + (x - bin.means[aggregateKey])/n
                    bin.powerSumAverage[aggregateKey] = bin.powerSumAverage[aggregateKey] + (x*x - bin.powerSumAverage[aggregateKey])/n
                    if (bin.ns[aggregateKey] - 1 > 0):
                        mean = bin.means[aggregateKey]
                        toSqrt = (bin.powerSumAverage[aggregateKey]*n - n*mean*mean)/(n - 1)
                        bin.sampleStandardDeviations[aggregateKey] = math.sqrt(max(0, toSqrt))
                    
                elif aggregateFunction == 'Count':
                   
                    currentValue = bin.count / progress
                
                bin.values[aggregateKey] = currentValue   
                bin.counts[aggregateKey] += 1.0          
    
    def __createAggregateKey(self, x):
        binIndex = []
        for idx, binRange in enumerate(self.binStructure.binRanges):
            value = x[self.dimensions[idx]]
            if isinstance(binRange, NominalBinRange):
                binIndex.append(self.binStructure.binRanges[idx].getIndexFromValue(value))    
            else:
                binIndex.append(self.binStructure.binRanges[idx].getIndex(value))
        
        return AggregateKey(tuple(binIndex))

        
    def __initializeBinStructure(self, binRanges):
        binStructure = BinStructure(binRanges)
        binStructure.createBins();
        return binStructure;
            
    def __createBinRanges(self, dataMins, dataMaxs, df):
        binRanges = []
        for idx, val in enumerate(self.dimensions):
            if self.dimensionAggregateFunctions[idx] != 'None':
                binRanges.append(AggregatedBinRange.initialize())
            elif df.dtypes[val] == 'float64' or df.dtypes[val] == 'int64':
                binRanges.append(QuantitativeBinRange.initialize(dataMins[idx], dataMaxs[idx], self.nrOfBins[idx], df.dtypes[val] == 'int64'))
            else:
                binRanges.append(NominalBinRange.initialize(df, val))
        return binRanges
 
 
 
        
class BinStructure():

    def __init__(self, binRanges):
        self.binRanges = binRanges     
        self.bins = {}
        self.nullCount = 0
        self.z = 1.96;
    
    def toJson(self):
        data = {
            'binRanges' : [br.__dict__ for br in self.binRanges],
            'bins' : {str(list(k)): v.toJson() for k, v in self.bins.iteritems()},
            'nullCount': self.nullCount
        }
        return data
    
    def createBins(self):
        self.__recursiveCreateBins(self.binRanges, [])
            
    def __recursiveCreateBins(self, previousBinRangesLeft, previousSpans):
        binRangesLeft = []
        binRangesLeft.extend(previousBinRangesLeft);
        
        binRange = binRangesLeft[0];
        del binRangesLeft[0]
        
        for idx, val in enumerate(binRange.getBins()):
            span = Span(val, val, idx)
            spans = []
            spans.extend(previousSpans)
            spans.append(span);
            if len(binRangesLeft) == 0:          
                bin = Bin(spans, tuple([s.index for s in spans]))
                self.bins[bin.binIndex] = bin
            else:
                self.__recursiveCreateBins(binRangesLeft, spans);
    
    def initializeBatchAggregation(self, dimensions, toAggregate, toAggregateFunctions, brushes, df):
        for brushIndex in brushes:
            for idx, aggregate in enumerate(toAggregate):
                aggregateFunction = toAggregateFunctions[idx]
                aggregateKey = (aggregate, aggregateFunction, brushIndex)
                dType = df.dtypes[aggregate]
                
                for bin in self.bins.values():
                    if not aggregateKey in bin.counts:
                        bin.counts[aggregateKey] = 0.0
                        
                        if not aggregateKey in bin.values:
                            if dType == 'float64' or dType == 'int64' or aggregateFunction == 'Count':
                                bin.values[aggregateKey] = 0.0
                            else:
                                bin.values[aggregateKey] = None
                                
                    if not aggregateKey in bin.margins:
                        bin.margins[aggregateKey] = 0
                        bin.marginsAbsolute[aggregateKey] = 0
                        bin.powerSumAverage[aggregateKey] = 0
                        bin.ns[aggregateKey] = 0
                        bin.sampleStandardDeviations[aggregateKey] = 0
                        bin.means[aggregateKey] = 0
                            
    def endBatchAggregation(self, dimensions, toAggregate, toAggregateFunctions, brushes, df, progress):   
        for brushIndex in brushes:
            for idx, aggregate in enumerate(toAggregate):
                aggregateFunction = toAggregateFunctions[idx]
                aggregateKey = (aggregate, aggregateFunction, brushIndex)
                dType = df.dtypes[aggregate]
                
                sumCount = sum([b.counts[aggregateKey] for b in self.bins.values()])
                for bin in self.bins.values():
                    bin.countsInterpolated[aggregateKey] = bin.counts[aggregateKey] / progress
                    totalCountInterpolated = bin.countsInterpolated[aggregateKey]
                    
                    if bin.counts[aggregateKey] > 1:
                        if aggregateFunction == 'Count':
                            toSqrt = (totalCountInterpolated - bin.counts[aggregateKey])/(totalCountInterpolated - 1)                        
                            fpc = math.sqrt(max(0, toSqrt))

                            probability = bin.counts[aggregateKey] / sumCount;
                            margin = ((probability*(1.0 - probability))/ sumCount) * fpc
                            margin = math.sqrt(margin) * self.z
                            if not math.isnan(margin) and not math.isnan(totalCountInterpolated):
                                bin.marginsAbsolute[aggregateKey] = margin*totalCountInterpolated
                                bin.margins[aggregateKey] = margin
                                
                        elif aggregateFunction == 'Avg':
                            fpc = math.sqrt((totalCountInterpolated - bin.counts[aggregateKey]) / (totalCountInterpolated - 1))
                            bin.margins[aggregateKey] = ((bin.sampleStandardDeviations[aggregateKey] / math.sqrt(bin.counts[aggregateKey])) * fpc)*self.z
                            bin.marginsAbsolute[aggregateKey] = ((bin.sampleStandardDeviations[aggregateKey] / math.sqrt(bin.counts[aggregateKey])) * fpc) * self.z
                    
    def aggregate(self, binIndexTuple, dimensions, toAggregate, toAggregateFunctions, brushes, df, row, dfDtypes, progress):
        maxCount = sys.float_info.max;
        bin = self.bins[binIndexTuple]
        bin.count += 1

        #row = df.loc[rowIndex]
        for brushIndex in brushes:
            for idx, aggregate in enumerate(toAggregate):
                aggregateFunction = toAggregateFunctions[idx]
                aggregateKey = (aggregate, aggregateFunction, brushIndex)
                dType = dfDtypes[aggregate]
                        
                currentValue = bin.values[aggregateKey]
                value = row[aggregate]
                
                if aggregateFunction == 'Max':
                    currentValue = max(currentValue, value)
                    
                elif aggregateFunction == 'Min':
                    currentValue = min(currentValue, value)    
                    
                elif aggregateFunction == 'Sum':
                    currentValue = currentValue + value
                    
                elif aggregateFunction == 'Avg':
                    currentValue = ((currentValue * bin.counts[aggregateKey]) + value) / (bin.counts[aggregateKey] + 1)
                    n = bin.ns[aggregateKey] + 1
                    bin.ns[aggregateKey] = n
                    x = value
                    bin.means[aggregateKey] = bin.means[aggregateKey] + (x - bin.means[aggregateKey])/n
                    bin.powerSumAverage[aggregateKey] = bin.powerSumAverage[aggregateKey] + (x*x - bin.powerSumAverage[aggregateKey])/n
                    if (bin.ns[aggregateKey] - 1 > 0):
                        mean = bin.means[aggregateKey]
                        toSqrt = (bin.powerSumAverage[aggregateKey]*n - n*mean*mean)/(n - 1)
                        bin.sampleStandardDeviations[aggregateKey] = math.sqrt(max(0, toSqrt))
                    
                elif aggregateFunction == 'Count':
                    currentValue = bin.count / progress
                
                bin.values[aggregateKey] = currentValue   
                bin.counts[aggregateKey] += 1.0   
                
    def map(self, binStructure):
        self.nullCount += binStructure.nullCount;
        
        for oldBinIndex in binStructure.bins.keys():
            oldBin = binStructure.bins[oldBinIndex];
            newBinIndex = [];

            for d, val in enumerate(self.binRanges):
                newBinIndex.append(self.binRanges[d].getIndex(oldBin.spans[d].min))

            newBin = self.bins[tuple(newBinIndex)];

            if newBin.containsBin(oldBin):
                newBin.map(oldBin)
        
class Bin():
    def __init__(self, spans, binIndex):
        self.spans = spans
        self.binIndex = binIndex
        self.count = 0
        self.counts = {}
        self.values = {}
        self.countsInterpolated = {}
        self.margins = {}
        self.marginsAbsolute = {}
        self.means = {}
        self.powerSumAverage = {}
        self.sampleStandardDeviations = {}
        self.ns = {}
    
    def toJson(self):
        data = {
            'counts' : self.__reHash(self.counts),
            'countsInterpolated' : self.__reHash(self.countsInterpolated),
            'values' : self.__reHash(self.values),
            'margins' : self.__reHash(self.margins),
            'marginsAbsolute' : self.__reHash(self.marginsAbsolute)            
        }
        return data
        
    def __reHash(self, d):
        newD = {}
        for k, v in d.iteritems():
            subD = newD
            for i in range(len(k)):
                subK = k[i]
                if i != len(k)-1:
                    if not subK in subD:
                        tempD = {}
                        subD[subK] = tempD
                    subD = subD[subK]
                else:
                    subD[subK] = v
        return newD
    
    def containsBin(self, bin):
        for d, val in enumerate(self.spans):
            if (not ((bin.spans[d].min >= self.spans[d].min or __aboutEqual(bin.spans[d].min, self.spans[d].min)) and
                     (bin.spans[d].max <= self.spans[d].max or __aboutEqual(bin.spans[d].max, self.spans[d].max)))):
                return False
        return True
    
    def __aboutEqual(x, y):
        epsilon = max(abs(x), abs(y)) * 1E-15;
        return abs(x - y) <= epsilon;
        
    def map(self, bin):
        self.binIndex = bin.binIndex
        self.count = bin.count
        self.spans = bin.spans
        self.counts = bin.counts
        self.values = bin.values
        self.countsInterpolated = bin.countsInterpolated
        self.margins = bin.margins
        self.marginsAbsolute = bin.marginsAbsolute
        self.means = bin.means
        self.powerSumAverage = bin.powerSumAverage
        self.sampleStandardDeviations = bin.sampleStandardDeviations
        self.ns = bin.ns

class AggregateKey():
    def __init__(self,index):
        self.index = index  
    def __eq__(self, other):
        return other and self.index == other.index

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return index
        
class Span():
    def __init__(self, min, max, index):
        self.min = min     
        self.max = max  
        self.index = index  
        
    def __str__(self):
        return str(self.__dict__)
    def __unicode__(self):
        return 'u'+str(self.__dict__)
    def __repr__(self):
        return str(self.__dict__)